### PR TITLE
Allow "drawtype:<lua drawtype name>" in getIds()

### PR DESCRIPTION
**Pull request by Thomas--S**
Thanks to @SmallJoker for the help.

Useful e.g. for connected nodeboxes.

ToDo:

- [x] Add documentation
- [x] Further testing
- [x] (Don't copy-paste `EnumString INodeDefManager::DrawTypes[]`from `s_node.cpp`. ~~I tried to do this, but I failed. This needs to be done by someone else later.~~ _Thanks to Krock's help, I managed to fix this._)

Replaces #5579.

@paramat Better? :wink:

### CODE REVIEW

**Comment by paramat** [comment]
Interesting, so this adds the ability to use drawtype:xxxxx anywhere we can currently use group:xxxxx?
Might be a good idea to stress-test using dreambuilder with it's 10,000 or whatever nodes.

**Comment by Thomas--S** [comment]
PR updated. Now all drawtypes should work correctly.

**Comment by paramat** [comment]
If what yo want is for connected nodeboxes to be able to connect according to drawtype, that should be done in connected nodebox code, this is an over-powered way to do that.
However, this PR should be considered a feature done for many reasons, can people describe usecases? How needed is this?

**Comment by Thomas--S** [comment]
PR updated (some bugfixes, better performance)

paramat, you asked me to test this with Dreambuilder: It works fine with it.

I strongly doubt that implementing this in the connected nodeboxes code will be better. This will only lead to more code fragmentation.

**Comment by paramat** [comment]
Well if your intent is for new connected nodebox behaviour, that's the place to do it, instead of adding a powerful new feature. That's why this PR is now not about your original intent and must be judged on multiple uses, it's too big a feature to do just for connected nodeboxes alone.

**Comment by Thomas--S** [comment]
As far as I can see, the `getIds()` function is used for three things:

- Connected Nodeboxes
- `find_node_near` stuff
- ABMs and LBMs

Crafting recipes (they also support the group:____ syntax) seem to be handled somewhere else.

I think the drawtype:____ thing makes sense for find_nodes_near and ABMs/LBMs, too. An Example: You want to replace all plantlike nodes under water (they look a bit silly) with something else (e.g. normal water).

**Comment by rubenwardy** [comment]
Are you interested on improving this?

I'm torn over whether this is worth including or is too specialised

**Comment by paramat** [comment]
:-1: For my reasons given here and in the previous PR.

**Comment by Thomas--S** [comment]
I'm still interested in this PR.

I would update it as soon as I have time for this. But I think it is kind of pointless to further care about this PR because of the disapproval.

@paramat: What is your suggestion to implement this feature?

**Comment by paramat** [comment]
It only has 1 disapproval, if it gets 2 approvals it might be merged.
The description needs to summarise what it does. 
For others' sake, please could you answer my question:

> Interesting, so this adds the ability to use drawtype:xxxxx anywhere we can currently use group:xxxxx?

**Comment by Thomas--S** [comment]
Sorry for the late answer, I was quite busy during the last few days.

> Interesting, so this adds the ability to use drawtype:xxxxx anywhere we can currently use group:xxxxx?

It adds the ability to use draytype:xxxxx anywhere where getIds() is used internally. This is (as already stated above):

- Connected nodeboxes
- `find_node*` functions
- ABMs and LBMs

The first usage is already tested, it seems to work well. I will test the other two usages soon.

If the core devs prefer only to change the behaviour for connected nodeboxes, I think this won't be difficult to achieve (by adding an optional second parameter to the `getIds()` method).

I will add the according documentation as soon as core devs have decided if they want to have the change like it is now, or if they prefer the nodebox-only solution.

Thanks for your efforts!

**Comment by paramat** [comment]
Thanks that's clearer.

File: src/nodedef.cpp [hunk 1]
```diff
@@ -1113,24 +1154,46 @@ bool CNodeDefManager::getIds(const std::string &name,
...
+
+    	std::unordered_map<std::string, std::vector<content_t>>::const_iterator
+    		i = m_group_to_items.find(group);
+    	if (i == m_group_to_items.end())
+    		return true;
+
+    	const std::vector<content_t> &items = i->second;
+    	result.insert(result.end(), items.begin(), items.end());
+    	//printf("getIds: %dus\n", t.stop());
+    	return true;
```
Comment (paramat) [inline src/nodedef.cpp]: Code block above has wrong indentation

File: src/nodedef.cpp [hunk 2]
```diff
@@ -1113,24 +1154,46 @@ bool CNodeDefManager::getIds(const std::string &name,
...
+           }
+
+    	std::unordered_map<int, std::vector<content_t>>::const_iterator
+    		i = m_drawtype_to_items.find(drawtype);
+    	if (i == m_drawtype_to_items.end())
+    		return true;
+
+    	const std::vector<content_t> &ids = i->second;
+    	result.insert(result.end(), ids.begin(), ids.end());
+    	return true;
```
Comment (paramat) [inline src/nodedef.cpp]: Code block above has wrong indentation

File: src/nodedef.cpp [hunk 3]
```diff
@@ -1113,24 +1154,46 @@ bool CNodeDefManager::getIds(const std::string &name,
...
+    	std::string drawtype_name = name.substr(9);
+
+    	const EnumString *esp = DrawTypes;
+           while(esp->str){
+           	if(drawtype_name == std::string(esp->str)){
+           		drawtype = esp->num;
+           		break;
+           	}
+           	esp++;
+           }
```
Comment (paramat) [inline src/nodedef.cpp]: Code block above has wrong indentation

Comment (Thomas--S) [reply src/nodedef.cpp]: Thanks for pointing these mistakes out.

**Comment by Thomas--S** [comment]
ABMs and LBMs are tested now, too, and they also seem to work fine.

paramat, you stated there may not be enough need for this feature. I can understand this point of view for the `find_node*()` and ABM/LBM usages, but for connected nodeboxes, there is a need for this. 

Examples:

- Traffic barriers and roadwork/construction fences of the Streets Mod (connecting to all drawtype:normal)
- Simulating the old xpanes behaviour (connecting to all "solid" drawtypes)
- ...

If you prefer to have this feature only for nodeboxes, what do you think about my suggestion with the optional second parameter for `getIds()`?

**Comment by paramat** [comment]
I misunderstood earlier in the thread, i'm neutral now. I can see this doesn't slow anything down if drawtype:xxxxx isn't used.

**Comment by Thomas--S** [comment]
This should be ready for review now.

File: src/nodedef.cpp [hunk 4]
```diff
@@ -41,6 +41,45 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+		{NDT_RAILLIKE, "raillike"},
+		{NDT_NODEBOX, "nodebox"},
+		{NDT_MESH, "mesh"},
+		{NDT_PLANTLIKE_ROOTED, "plantlike_rooted"},
+		{0, NULL},
+	};
+
+bool INodeDefManager::string_to_enum (const EnumString *spec, int &result, const std::string &drawtype) const
+{
+	const EnumString *esp = spec;
```
Comment (SmallJoker) [inline src/nodedef.cpp]: Useless variable. Use `spec` instead of `esp`

File: src/nodedef.cpp [hunk 5]
```diff
@@ -41,6 +41,45 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+		{NDT_FIRELIKE, "firelike"},
+		{NDT_FENCELIKE, "fencelike"},
+		{NDT_RAILLIKE, "raillike"},
+		{NDT_NODEBOX, "nodebox"},
+		{NDT_MESH, "mesh"},
+		{NDT_PLANTLIKE_ROOTED, "plantlike_rooted"},
+		{0, NULL},
+	};
+
+bool INodeDefManager::string_to_enum (const EnumString *spec, int &result, const std::string &drawtype) const
```
Comment (SmallJoker) [inline src/nodedef.cpp]: Line too long, superfluous space near `(`
This function should be named `stringToEnum` or similar
Low priority: Could you add this function as `inline` into `c_types.h`? It's a helper function for this struct array.

File: src/nodedef.cpp [hunk 6]
```diff
@@ -41,6 +41,45 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+{
+	const EnumString *esp = spec;
+	while(esp->str) {
+		if(drawtype == std::string(esp->str)) {
```
Comment (SmallJoker) [inline src/nodedef.cpp]: Missing spaces after `while` and `if`

File: src/nodedef.cpp [hunk 7]
```diff
@@ -970,6 +1009,10 @@ class CNodeDefManager: public IWritableNodeDefManager {
 	// Note: Not serialized.
 	std::unordered_map<std::string, std::vector<content_t>> m_group_to_items;
 
+	// A mapping from drawtypes to a list of content_ts that belong to it.
+	// Note: Not serialized
+	std::unordered_map<int, std::vector<content_t> > m_drawtype_to_items;
```
Comment (SmallJoker) [inline src/nodedef.cpp]: Space after `content_t>` can be omitted.

**Comment by Thomas--S** [comment]
PR updated according to the comments. It would be good someone could check if the "inline" thing in c_types.h is coded correctly, as my C++ knowledge is very limited.

**Comment by rubenwardy** [comment]
I don't think this is a good idea, this whole PR feels odd to me. Just because something is drawn like something doesn't mean it is something. For example, signlike is used by more than signs.

**Comment by Thomas--S** [comment]
> I don't think this is a good idea, this whole PR feels odd to me. Just because something is drawn like something doesn't mean it is something. For example, signlike is used by more than signs.

This is certainly true for some drawtypes, but you can assume that e.g. a drawtype:normal node fills up the whole area `{-0.5, -0.5, -0.5, 0.5, 0.5, 0.5}`. This is often a good condition for letting a node connect to this.

> It's also possible to do this with groups already

I don't think that this is possible. Assume a mod wants to define a node with a connected nodebox, e.g. some kind of fence. This fence should connect to all "normal" nodes. When using groups, this mod would need to depend on every single other mod (impossible to do) and then redefine all node definitions to add the group.

This may seem like a rarely useful feature, but when developing the streets mod, there was often the need for this feature. Just think of a traffic barrier: It would be useful if this node could connect to every "normal" block, so that it connects e.g. to the walls of houses.

I hope this explanation was understandable, please excuse grammer and spelling mistakes. Thanks!

**Comment by rubenwardy** [comment]
> I don't think that this is possible.

With no dependencies:

```lua
local function addGroup(def)
    def.groups = table.copy(def.groups)
    def.groups.normal = 1
end

for name, def in pairs(minetest.registered_nodes) do
    if def.drawtype == "normal" then
        addGroup(def)
        minetest.override_item(name, def)
    end
end

local old_reg = minetest.register_node
function minetest.register_node(name, def)
    if def.drawtype == "normal" then
        addGroup(def)
    end
    old_reg(name, def)
end

**Comment by Thomas--S** [comment]
Thanks rubenwardy, for sharing this piece of code. I didn't know this was possible. I will leave this open, in case there is a core dev who likes this PR, but I think this is kind of low priority then.

**Comment by paramat** [comment]
:-1: I don't see enough need to add such a big change.
Instead could a custom mod node be used which has a group that allows a connected nodebox to connect to it?

**Comment by Thomas--S** [comment]
@paramat It's okay for me when you close this. I can understand that this feature will only be needed very rarely. rubenwardy showed me a good solution how to achieve the same functionality mod-side. This is enough for me. I only left this open in case anybody else is interested in this, but it doesn't seem so.

So, feel free to close this, when you like.

Paramat, rubenwardy and SmallJoker: I really appreciate your efforts in reviewing this and helping me to find a mod-side solution. Thanks! :+1:

**Comment by paramat** [comment]
Ok, it seems unpopular with devs so i'll close.

**Pull request closed without merging**